### ISSUE TITLE
Texture Brightness High Exposure

### ISSUE THREAD

**Issue by Habsburgg** [cdcseacave/openMVS#1020]
Hi, for texture maps there are many area luminosity highly exposed. Is there a suitable way to reduce this phenomenon? Or is there any way to lower the weight of this region, giving priority to non-highly exposed image regions.
Looking forward to your reply.
![111](https://github.com/cdcseacave/openMVS/assets/56535633/158714c0-ef14-4080-a000-6e9e87415190)

**Comment by cdcseacave** [comment]
are these highly exposed regions in the source images?

**Comment by Habsburgg** [comment]
> are these highly exposed regions in the source images?

Yes, but there are many pictures from different angles corresponding to the same place on the model, some pictures are high exposure, some are not.

**Comment by cdcseacave** [comment]
1. for best 3D reconstruction results (not only texture) images should be acquired at good, but diffuse light, and equal or close to equal exposure (constant light).
2. since 1 is a requirement, OpenMVS does not have a good way now to select images without glossiness artifacts, in fact as it is now it is biassed to select such images
3. help is needed to implement a better texturing algorithm, pls give a sign if interested

**Comment by Habsburgg** [comment]
thank you for your reply,
1. I marked the high-exposure areas in the image as much as possible, and reduced the scores of these pixels in the view selection stage, but the effect was not good. If I have time, I will continue to study.
2. I encountered a problem. There is still a color difference between the texturePatches. I saw that you used Poisson fusion. Why is there such a big difference? Can opencv's seamlessClone replace this function?
![textureBorder](https://github.com/cdcseacave/openMVS/assets/56535633/2089372c-93fe-4039-aac1-f00e2b634f40)
![texture](https://github.com/cdcseacave/openMVS/assets/56535633/6ece6bea-6963-47ed-929a-eb0389ac3836)

**Comment by cdcseacave** [comment]
good suggestion, a different blending procedure might be better, pls try

**Comment by Habsburgg** [comment]
I did an experiment. Set the color of the border of the texture block to black as the boundary condition of the Poisson calculation, and the texture map result is shown in the figure. Does this mean that there is a problem with setting the edge color of the texture block in LocalSeamLeveling, causing the Poisson calculation to be wrong.
I'm not sure if my guess is correct.
![1111](https://github.com/cdcseacave/openMVS/assets/56535633/19274144-14d4-4d9a-a0dc-ac7d1d6e6504)
![2222](https://github.com/cdcseacave/openMVS/assets/56535633/d87320a4-97ec-4de5-a843-75847c2bc316)

**Comment by cdcseacave** [comment]
Interesting experiment. However not clear to me what is the results before and after, and which edge did you set to black exactly. Can you add more screenshots?

**Comment by Habsburgg** [comment]
openMVS:
![11](https://github.com/cdcseacave/openMVS/assets/56535633/f588e38a-ea8c-4f85-adba-68594908d762)
![12](https://github.com/cdcseacave/openMVS/assets/56535633/45d95c4b-54a5-4b8a-bfeb-13ebe6859f04)

**Comment by Habsburgg** [comment]
mvs-texturing:
![21](https://github.com/cdcseacave/openMVS/assets/56535633/7de9738c-ca58-4784-a174-f7eb607665ad)
![22](https://github.com/cdcseacave/openMVS/assets/56535633/92cb394f-4a12-4651-9605-3a29b1cc49d6)

**Comment by Habsburgg** [comment]
> Interesting experiment. However not clear to me what is the results before and after, and which edge did you set to black exactly. Can you add more screenshots?

I set the junction of different textures to black, and the original code is set to the average color.
By comparison, I suspect that openmvs misses some border pixels, doesn't change their color, causing the poisson calculation to still use the original color, which leads to weird transition areas between different textures.

**Comment by Habsburgg** [comment]
This is the mask image I saved, and you can see that some edges are not set to gray "border".
In the DrawLine function, "Bresenham's line algorithm" is used to calculate the texture edge pixels. Is there any problem here? Looking forward to your reply.
![mask2](https://github.com/cdcseacave/openMVS/assets/56535633/2bea2535-8630-4c25-acc3-c5942d2cb74c)
![mask1](https://github.com/cdcseacave/openMVS/assets/56535633/9084b86e-f9d8-44ee-a0be-f52ebca50661)

**Comment by cdcseacave** [comment]
thanks for the detailed experiment
a couple of thoughts:
- mvs-texturing is using [mapmap](https://github.com/dthuerck/mapmap_cpu) instead of LBP to assign a image to each face, which should be not only faster, but better/fewer patches, and less seams
- recently I discovered a bug in how the camera matrix K is scaled, and I think the same bug affects also other parts of OpenMVS where the image coordinates are scaled, possibly also in texturing, though I can not see one at a quick look
most probably though might be a problem with drawing the line, like you say, or something related; it would be of great help if you could look into this and fix it; you could compare on how this is done in mvs-texturing

**Comment by cdcseacave** [comment]
Brightness/exposure differences between views are handled by the global color adjustment in TextureMesh (Poisson blending across texture seams). If brightness variation persists, ensure consistent exposure during capture.
